php/class-settings.php

```diff
@@ -0,0 +1,148 @@
+	}
+
+	/**
+	 * Register any hooks that this component needs.
+	 */
+	public function register_hooks() {
+		add_action( 'admin_menu', array( $this, 'add_submenu_pages' ) );
+		add_action( 'admin_init', array( $this, 'register_settings' ) );
+		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+		add_action( 'admin_notices', array( $this, 'show_notices' ) );
+	}
+
+	/**
+	 * Enqueue scripts and styles used by the Settings screen.
+	 *
+	 * @return void
+	 */
+	public function enqueue_scripts() {
+		$page = filter_input( INPUT_GET, 'page', FILTER_SANITIZE_STRING );
+
+		// Enqueue scripts and styles on the edit screen of the Block post type.
+		if ( $this->slug === $page ) {
+			wp_enqueue_style(
+				$this->slug,
+				$this->plugin->get_url( 'css/admin.settings.css' ),
+				array(),
+				$this->plugin->get_version()
+			);
+		}
+	}
+
+	/**
+	 * Add submenu pages to the Block Lab menu.
+	 */
+	public function add_submenu_pages() {
+		add_submenu_page(
+			'edit.php?post_type=block_lab',
+			__( 'Block Lab Settings', 'block-lab' ),
+			__( 'Settings', 'block-lab' ),
+			'manage_options',
+			$this->slug,
+			array( $this, 'render_page' )
+		);
+	}
+
+	/**
+	 * Register Block Lab settings.
+	 */
+	public function register_settings() {
+		register_setting( 'block-lab-license-key', 'block_lab_license_key' );
+	}
+
+	/**
+	 * Render the Settings page.
+	 */
+	public function render_page() {
+		$tab = filter_input( INPUT_GET, 'tab', FILTER_SANITIZE_STRING );
+
+		ob_start();
+
+		$this->render_page_header();
+
+		switch ( $tab ) {
+			case 'license':
+			default:
+				include 'views/license.php';
+				break;
+		}
+
+		$html = ob_get_clean();
+		echo '<div class="wrap block-lab-settings">' . $html . '</div>'; // phpcs: XSS okay.
+	}
+
+	/**
+	 * Render the Settings page header.
+	 */
+	public function render_page_header() {
+		?>
+		<h2><?php echo esc_html( get_admin_page_title() ); ?></h2>
+		<h2 class="nav-tab-wrapper">
+			<a href="<?php echo esc_url( add_query_arg( 'tab', 'license' ) ); ?>" title="License" class="nav-tab nav-tab-active dashicons-before dashicons-nametag">
+				<?php esc_html_e( 'License' ); ?>
+			</a>
+			<a href="https://github.com/getblocklab/block-lab/wiki" target="_blank" class="nav-tab dashicons-before dashicons-info">
+				<?php esc_html_e( 'Documentation' ); ?>
+			</a>
+			<a href="https://wordpress.org/support/plugin/block-lab/" target="_blank" class="nav-tab dashicons-before dashicons-sos">
+				<?php esc_html_e( 'Help' ); ?>
+			</a>
+		</h2>
+		<?php
+	}
+
+	/**
+	 * Prepare notices to be displayed after saving the settings.
+	 *
+	 * @param string $notice The notice text to display.
+	 */
+	public function prepare_notice( $notice ) {
+		$notices   = get_option( 'block_lab_notices', array() );
+		$notices[] = $notice;
+		update_option( 'block_lab_notices', $notices );
+	}
+
+	/**
+	 * Show any admin notices after saving the settings.
+	 */
+	public function show_notices() {
```

Comment:
This is an interesting way of queuing the notices to display 😄

Reply:
Yes, this happens this way for a reason. I am using WordPress' Settings API to save all the options, which means I don't have to write any code which checks for $_POST variables or any code to `update_option` s.

The side effect is that the form goes off to `options.php` to save, and then redirects back to the settings page. After the redirect, I've got no $_POST variables to speak of, so I don't know  if the form was saved, or if it was saved successfully.

Two ways I see around this:
- Hijack the options saving process and redirect back to the settings page with an additional $_GET parameter indicating the result (then using that result to display a notice)
- Save the notice in the database, and check if it exists when the page loads (and show if it does)

I opted for the second method, since I feel it's cleaner, and potentially reusable in other situations.

Reply:
Sounds good, thanks for explaining this 😄